### PR TITLE
JS Objects in Knobs addon

### PR DESCRIPTION
**Pull request by z4o4z**
Issue:
#1187 #1186

## What I did
Added JS Objects support to Knobs addon
Added react-ace as code editor 

## How to test
`npm run storybook` - look at 'with object' story

### CODE REVIEW

**Comment by shilman** [comment]
@z4o4z how big is the ace-editor/brace dependency?

**Comment by z4o4z** [comment]
@shilman Hi, all new libs(react-ace, brace, brace/mode/javascript, brace/theme/github) have size about 710kb (not gzip)

**Review by danielduan — APPROVED** [review]
Love the ability to read a knobs object. it's been really annoying seeing `[object],[object]`

I'll leave it up to others whether the editor is worth adding to our codebase.

File: addons/knobs/src/components/types/Object.js
Comment (z4o4z) [inline addons/knobs/src/components/types/Object.js]: Hi @danielduan . I didn't use json.stringify because I tried to add support for functions. In future we can add support for other JS features(regexp, Map, Set, and etc)

Comment (danielduan) [inline addons/knobs/src/components/types/Object.js]: makes sense. thanks @z4o4z

**Comment by shilman** [comment]
@tmeasday @ndelangen any thoughts on this re: code bloat with the new editor?

**Comment by shilman** [comment]
@z4o4z Finally had a chance to test this out. Love ❤️  the Object support and the function support is really funky and cool.

I also think:
1. Examples should be added to `examples/cra-kitchen-sink` so we can see how the different addons interact in a single spot (which we'll run automated tests against soon).
2. We should either get rid of the addon-specific example OR add lerna-support for them somehow (cc @ndelangen)
3. We should split the `react-ace` work into a separate PR and not merge it until we have a better use case for it. For example @wcastand was working on support for a full code editor addon. (I am very concerned about Storybook getting bigger and slower since 3.0, and am very sensitive to taking on big dependencies, and think we're probably overdue for a performance milestone soon.)

What do you think? I'm happy to do some of this work, or load-balance.

**Comment by shilman** [comment]
For @ndelangen 

storybook

storybook

**Comment by z4o4z** [comment]
Hi @shilman, Excellent suggestions! I will add demos to `examples/cra-kitchen-sink`.
And can you help me with it?
>  split the react-ace work into a separate PR

**Comment by ndelangen** [comment]
minor linting error:
```
/tmp/storybook/addons/knobs/src/components/types/Object.js
  56:14  error    Arrow function should not return assignment               no-return-assign
```

**Comment by ndelangen** [comment]
@z4o4z Can you take another look at this? I think the snapshots are outdated!

**Comment by z4o4z** [comment]
@ndelangen yep

File: addons/knobs/src/components/types/Object.js
Comment (Hypnosphi) [inline addons/knobs/src/components/types/Object.js]: We need some XSS protection here

Comment (danielduan) [reply addons/knobs/src/components/types/Object.js]: Are we afraid of users scripting themselves or a knob url link with something potentially malicious inside it?

I think we can get around this by not allowing the javascript to be serialized into the URL if it's a real concern.

Comment (Hypnosphi) [reply addons/knobs/src/components/types/Object.js]: Yeah, mostly the latter

**Comment by danielduan** [comment]
@z4o4z do you think you can remove the js object serialization so we can move this PR forward?

File: addons/knobs/src/components/types/Object.js
Comment (ndelangen) [reply addons/knobs/src/components/types/Object.js]: 🎖 for you @Hypnosphi

Comment (Hypnosphi) [reply addons/knobs/src/components/types/Object.js]: Actually it shouldn't be deserialized from URL, either

**Review by Hypnosphi — CHANGES_REQUESTED** [review]
See https://github.com/storybooks/storybook/pull/1460#discussion_r150982016

**Pull request closed without merging**
